fog/ledger/enclave/impl/src/lib.rs

```diff
@@ -191,6 +196,164 @@ where
 
         Ok(())
     }
+
+    fn ledger_store_init(&self, ledger_store_id: ResponderId) -> Result<NonceAuthRequest> {
+        Ok(self.ake.backend_init(ledger_store_id)?)
+    }
+
+    #[allow(unused_variables)]
+    fn ledger_store_connect(
+        &self,
+        ledger_store_id: ResponderId,
+        ledger_store_auth_response: NonceAuthResponse,
+    ) -> Result<()> {
+        Ok(self
+            .ake
+            .backend_connect(ledger_store_id, ledger_store_auth_response)?)
+    }
+
+    fn decrypt_and_seal_query(
+        &self,
+        client_query: EnclaveMessage<ClientSession>,
+    ) -> Result<SealedClientMessage> {
+        Ok(self.ake.decrypt_client_message_for_enclave(client_query)?)
+    }
+
+    fn create_multi_key_image_store_query_data(
+        &self,
+        sealed_query: SealedClientMessage,
+    ) -> Result<Vec<EnclaveMessage<NonceSession>>> {
+        Ok(self
+            .ake
+            .reencrypt_sealed_message_for_backends(&sealed_query)?)
+    }
+
+    fn collate_shard_query_responses(
+        &self,
+        sealed_query: SealedClientMessage,
+        shard_query_responses: BTreeMap<ResponderId, EnclaveMessage<NonceSession>>,
+    ) -> Result<EnclaveMessage<ClientSession>> {
+        if shard_query_responses.is_empty() {
+            return Ok(EnclaveMessage::default());
+        }
+        let channel_id = sealed_query.channel_id.clone();
+        let client_query_plaintext = self.ake.unseal(&sealed_query)?;
+        // TODO this will (possibly?) be used when we implement obliviousness
+        let _client_query_request: CheckKeyImagesRequest =
+            mc_util_serial::decode(&client_query_plaintext).map_err(|e| {
+                log::error!(self.logger, "Could not decode client query request: {}", e);
+                Error::ProstDecode
+            })?;
+
+        let shard_query_responses = shard_query_responses
+            .into_iter()
+            .map(|(responder_id, query_response)| {
+                let plaintext_bytes = self.ake.backend_decrypt(&responder_id, &query_response)?;
+                let query_response: CheckKeyImagesResponse =
+                    mc_util_serial::decode(&plaintext_bytes)?;
+
+                Ok(query_response)
+            })
+            .collect::<Result<Vec<CheckKeyImagesResponse>>>()?;
+
+        // NOTES:
+        // num_blocks = min(responses.num_blocks)
+        // global_txo_count = min(global_txo_count) TODO CONFIRM
+        // results = cat(responses.results)
+        // latest_block_version = max(responses.latest_block_version)
+        // max_block_version = max(latest_block_version,
+        // mc_transaction_core::MAX_BLOCK_VERSION
+
+        // TODO no unwraps
```

Comment:
nit: Do the unwraps get removed in a follow-up PR? I'd recommend putting it in this PR, as that's something that can easily be forgetten.

Reply:
This entire method gets moved into an enclave method which obliviously collates the responses from the stores later on. That PR isn't in the stack yet though. If you like I can pull that into this PR as well, it'll just be bigger.

Reply:
Resolved in PR #2913 .